src/util.js

```diff
@@ -276,6 +276,45 @@ function copyDescriptors(from, to, filter) {
     }
 }
 
+var _hookIdCounter = 0;
+// This function should not be used for methods that return a value
+// Using this function implies small performance overhead, should be used only
```

Comment:
You are right, the benchmark shows, that "implementations loop" method is much worse than "function wrapping", will change it. Attaching benchmark code, FYI, it can be run as is browser console.

```javascript
var objectKeys = null;
if (this === undefined) {
    objectKeys = Object.keys;
} else {
    var has = {}.hasOwnProperty;
	objectKeys = function (o) {
        var ret = [];
        for (var key in o) {
            if (has.call(o, key)) {
                ret.push(key);
            }
        }
        return ret;
    };
}

function hookToArrayImpl(prototypeObject, methodName, extension) {
    var existingMethodImpl = prototypeObject[methodName];
    if (typeof existingMethodImpl === "function" &&
        typeof extension === "function") {
        if (!existingMethodImpl.extensions) {
            throw new Error("Trying to extend overriden method,"+
                " please use util.js:hookTo() to extend it");
        }
        extension._hookExtensionId = ++_hookIdCounter;
        existingMethodImpl.extensions[extension._hookExtensionId] = extension;
    } else if (typeof existingMethodImpl === "undefined") {
        prototypeObject[methodName] = function () {
            var extensions = prototypeObject[methodName].extensions;
            var keys = objectKeys(extensions);
            for (var i = 0; i < keys.length; i++) {
                extensions[keys[i]].apply(this, arguments);
            }
        };
        // Using object instead of array of functions to allow optimizations
        prototypeObject[methodName].extensions = {};
        if (typeof extension === "function") {
            extension._hookExtensionId = ++_hookIdCounter;
            prototypeObject[methodName].extensions[_hookIdCounter] = extension;
        }
    } else {
        throw new Error("Trying to wrap " + typeof existingMethodImpl +
            ", expecting a function or undefined");
    }
}

function hookToWrapImpl(prototypeObject, methodName, extension) {
    var existingMethodImpl = prototypeObject[methodName];
    if (typeof existingMethodImpl === "function") {
        prototypeObject[methodName] = function () {
            existingMethodImpl.apply(this, arguments);
            extension.apply(this, arguments);
        };
    } else if (typeof existingMethodImpl === "undefined") {
        prototypeObject[methodName] = extension;
    } else {
        throw new Error("Trying to wrap " + typeof existingMethodImpl + 
            ", expecting a function or undefined");
    }
}

function emptyFunction () {};
function smallFunction () { window.a = Math.random() };
function mediumFunction () { window.a = Math.random() * Math.random()};
function largeFunction () { for (var i=0; i<100; i++) {window.a += Math.random()}};


function runBenchmark(hookToImpl, extensions) {
    var Class = function() {};
	for (var i=0; i<extensions.length; i++) {
	    hookToImpl.call(null, Class.prototype, "method", extensions[i]);
    }
	var object = new Class();
	var start = new Date();
    for (var i=0; i<100000; i++) {
        object.method();
    }
	return  new Date() - start;
}

// Most common use case
console.log("hookToArrayImpl, standard use case: " + runBenchmark(hookToArrayImpl, [emptyFunction,smallFunction]));
console.log("hookToWrapImpl, standard use case: " + runBenchmark(hookToWrapImpl, [emptyFunction,smallFunction]));

console.log("hookToArrayImpl, heavy use case: " + runBenchmark(hookToArrayImpl, [emptyFunction,smallFunction,mediumFunction,largeFunction]));
console.log("hookToWrapImpl, heavy use case: " + runBenchmark(hookToWrapImpl, [emptyFunction,smallFunction,mediumFunction,largeFunction]));
```